File: plugins/codemodder-plugin-maven/src/main/java/io/codemodder/plugins/maven/operator/SimpleInsert.java
package io.codemodder.plugins.maven.operator;

import java.util.List;
import org.dom4j.Element;
import org.dom4j.Node;

/**
 * Represents a POM Upgrade Strategy by simply adding a dependency/ section (and optionally a
 * dependencyManagement/ section as well)
 */
class SimpleInsert implements Command {

  private final boolean validateDependencyExistence;

  /**
   * @param validateDependencyExistence is true when we only want to this command alone to insert
   *     dependencies in a CommandChain for dependency existence validations
   */
  SimpleInsert(final boolean validateDependencyExistence) {
    this.validateDependencyExistence = validateDependencyExistence;
  }

  /**
   * Executes the POM upgrade strategy by adding a dependency and optionally a dependencyManagement
   * section.
   *
   * @param pm The ProjectModel containing the configuration and settings for the upgrade.
   * @return `true` if the upgrade is successful, `false` otherwise.
   */
  @Override
  public boolean execute(ProjectModel pm) {

    if (validateDependencyExistence && checkDependencyExists(pm)) {
      return true;
    }

    List<Node> dependencyManagementNodeList =
        Util.selectXPathNodes(pm.getPomFile().getResultPom(), "/m:project/m:dependencyManagement");

    Element dependenciesNode;
    if (dependencyManagementNodeList.isEmpty()) {
      Element newDependencyManagementNode =
          Util.addIndentedElement(
              pm.getPomFile().getResultPom().getRootElement(),
              pm.getPomFile(),
              "dependencyManagement");

      dependenciesNode =
          Util.addIndentedElement(newDependencyManagementNode, pm.getPomFile(), "dependencies");
    } else {
      dependenciesNode = ((Element) dependencyManagementNodeList.get(0)).element("dependencies");
    }

    Element dependencyNode = appendCoordinates(dependenciesNode, pm);

    Element versionNode = Util.addIndentedElement(dependencyNode, pm.getPomFile(), "version");

    Util.upgradeVersionNode(pm, versionNode, pm.getPomFile());

    List<Node> dependenciesNodeList =
        Util.selectXPathNodes(pm.getPomFile().getResultPom(), "//m:project/m:dependencies");

    Element rootDependencyNode;
    if (dependenciesNodeList.isEmpty()) {
      rootDependencyNode =
          Util.addIndentedElement(
              pm.getPomFile().getResultPom().getRootElement(), pm.getPomFile(), "dependencies");
    } else if (dependenciesNodeList.size() == 1) {
      rootDependencyNode = (Element) dependenciesNodeList.get(0);
    } else {
      throw new IllegalStateException("More than one dependencies node");
    }

    appendCoordinates(rootDependencyNode, pm);

    return true;
  }

  private boolean checkDependencyExists(final ProjectModel pm) {
    String lookupExpressionForDependency =
        Util.buildLookupExpressionForDependency(pm.getDependency());

    List<Node> matchedDependencies =
        Util.selectXPathNodes(pm.getPomFile().getResultPom(), lookupExpressionForDependency);

    return !matchedDependencies.isEmpty();
  }

  /** Creates the XML Elements for a given dependency */
  private Element appendCoordinates(Element dependenciesNode, ProjectModel c) {
    Element dependencyNode =
        Util.addIndentedElement(dependenciesNode, c.getPomFile(), "dependency");

    Element groupIdNode = Util.addIndentedElement(dependencyNode, c.getPomFile(), "groupId");

    Dependency dep = c.getDependency();
    if (dep != null) {
      groupIdNode.setText(dep.getGroupId());
    }

    Element artifactIdNode = Util.addIndentedElement(dependencyNode, c.getPomFile(), "artifactId");

    if (dep != null) {
      artifactIdNode.setText(dep.getArtifactId());
    }

    return dependencyNode;
  }

  /**
   * This method is not used for this strategy and always returns `false`.
   *
   * @param c The ProjectModel.
   * @return `false`.
   */
  @Override
  public boolean postProcess(ProjectModel c) {
    return false;
  }
}

File: pom-operator/src/main/java/io/github/pixee/maven/operator/SimpleInsert.kt
package io.github.pixee.maven.operator

import io.github.pixee.maven.operator.Util.addIndentedElement
import io.github.pixee.maven.operator.Util.selectXPathNodes
import io.github.pixee.maven.operator.Util.upgradeVersionNode
import org.dom4j.Element

/**
 * Represents a POM Upgrade Strategy by simply adding a dependency/ section (and optionally a dependencyManagement/ section as well)
 */
val SimpleInsert = object : Command {
    override fun execute(pm: ProjectModel): Boolean {
        val dependencyManagementNodeList =
            pm.pomFile.resultPom.selectXPathNodes("/m:project/m:dependencyManagement")

        val dependenciesNode = if (dependencyManagementNodeList.isEmpty()) {
            val newDependencyManagementNode =
                pm.pomFile.resultPom.rootElement.addIndentedElement(
                    pm.pomFile,
                    "dependencyManagement"
                )

            val dependencyManagementNode =
                newDependencyManagementNode.addIndentedElement(pm.pomFile, "dependencies")

            dependencyManagementNode
        } else {
            (dependencyManagementNodeList.first() as Element).element("dependencies")
        }

        val dependencyNode = appendCoordinates(dependenciesNode, pm)

        val versionNode = dependencyNode.addIndentedElement(pm.pomFile, "version")

        upgradeVersionNode(pm, versionNode, pm.pomFile)

        val dependenciesNodeList =
            pm.pomFile.resultPom.selectXPathNodes("//m:project/m:dependencies")

        val rootDependencyNode: Element = if (dependenciesNodeList.isEmpty()) {
            pm.pomFile.resultPom.rootElement.addIndentedElement(pm.pomFile, "dependencies")
        } else if (dependenciesNodeList.size == 1) {
            dependenciesNodeList[0] as Element
        } else {
            throw IllegalStateException("More than one dependencies node")
        }

        appendCoordinates(rootDependencyNode, pm)

        return true
    }

    /**
     * Creates the XML Elements for a given dependency
     */
    private fun appendCoordinates(
        dependenciesNode: Element,
        c: ProjectModel
    ): Element {
        val dependencyNode = dependenciesNode.addIndentedElement(c.pomFile, "dependency")

        val groupIdNode = dependencyNode.addIndentedElement(c.pomFile, "groupId")

        val dep = c.dependency!!

        groupIdNode.text = dep.groupId

        val artifactIdNode = dependencyNode.addIndentedElement(c.pomFile, "artifactId")

        artifactIdNode.text = dep.artifactId

        return dependencyNode
    }

    override fun postProcess(c: ProjectModel): Boolean = false
}

